Add QMK Tap Dance Feature to ESC/Backtick key

In #3 we swapped `Esc` and Backtick (`` ` ``), which overall is a win.

However, it does make `Esc` slightly less accessible (now `Fn` + `Esc` is necessary).

I was reading about the [QMK Tap Dance](https://beta.docs.qmk.fm/using-qmk/software-features/feature_tap_dance) feature and if I understand correctly, we could keep the Esc key as Backtick (`` ` ``) _but_ when the Esc key is double-tapped, it would act as `Esc` (instead of backtick).

Note: When defining the keycode, e.g.

```
enum tapdance_keycodes {
    TD_ESCBT,              //Tap Dance Esc / Backtick
};
```

we can NOT add it to the existing `alt_keycodes` block

```
// Do NOT add tapdance keycode here!
enum alt_keycodes {
    U_T_AUTO = SAFE_RANGE, //USB Extra Port Toggle Auto Detect / Always Active
    U_T_AGCR,              //USB Toggle Automatic GCR control
    DBG_TOG,               //DEBUG Toggle On / Off
    DBG_MTRX,              //DEBUG Toggle Matrix Prints
    DBG_KBD,               //DEBUG Toggle Keyboard Prints
    DBG_MOU,               //DEBUG Toggle Mouse Prints
    MD_BOOT,               //Restart into bootloader after hold timeout
};
```

when I do, I'm seeing error messages like

```
Compiling: keyboards/massdrop/alt/keymaps/salcode-drop-alt/keymap.c                                 [OK]
Linking: .build/massdrop_alt_salcode-drop-alt.elf                                                   [ERRORS]
 |
 | /usr/local/Cellar/arm-gcc-bin@8/8-2019-q3-update_1/bin/../lib/gcc/arm-none-eabi/8.3.1/../../../../arm-none-eabi/bin/ld: .build/massdrop_alt_salcode-drop-alt.elf section `.relocate' will not fit in region `ram'
 | /usr/local/Cellar/arm-gcc-bin@8/8-2019-q3-update_1/bin/../lib/gcc/arm-none-eabi/8.3.1/../../../../arm-none-eabi/bin/ld: region `ram' overflowed by 679208 bytes
 | collect2: error: ld returned 1 exit status
 |
make[1]: *** [.build/massdrop_alt_salcode-drop-alt.elf] Error 1
make: *** [massdrop/alt:salcode-drop-alt] Error 1
Make finished with errors
```

for more details see this reddit post [QMK] region `ram' overflowed](https://www.reddit.com/r/olkb/comments/bfj442/qmk_region_ram_overflowed/)

--- a/CHANGELOG.md
+++ b/CHANGELOG.md
@@ -5,6 +5,11 @@ All notable changes to this project will be documented in this file.
 ### Changed
 - Swap left `Alt` and left `GUI` keys. Right `Alt` remains unchanged. (#1, #6)
 - Swap `Esc` and Backtick (`` ` ``) (#3)
+- Add [QMK Tap Dance Feature](https://beta.docs.qmk.fm/using-qmk/software-features/feature_tap_dance) to ESC/Backtick key, so the behavior when hitting the key varies based on the number of times it is hit in rapid succession (#5):
+    - Once: `` ` ``
+    - Twice: `<Esc>`
+    - Three times: `` ``` ``
+    - etc.
 
 ### Added
 - Add default `massdrop/alt` configuration

--- a/keymap.c
+++ b/keymap.c
@@ -38,7 +38,7 @@ qk_tap_dance_action_t tap_dance_actions[] = {
 
 const uint16_t PROGMEM keymaps[][MATRIX_ROWS][MATRIX_COLS] = {
     [0] = LAYOUT_65_ansi_blocker(
-        KC_GRV,                                                                                                                                \
+        TD(TD_BACKTICK_ESC),                                                                                                                   \
                  KC_1,    KC_2,    KC_3,    KC_4,    KC_5,    KC_6,    KC_7,    KC_8,    KC_9,    KC_0,    KC_MINS, KC_EQL,  KC_BSPC, KC_DEL,  \
         KC_TAB,  KC_Q,    KC_W,    KC_E,    KC_R,    KC_T,    KC_Y,    KC_U,    KC_I,    KC_O,    KC_P,    KC_LBRC, KC_RBRC, KC_BSLS, KC_HOME, \
         KC_CAPS, KC_A,    KC_S,    KC_D,    KC_F,    KC_G,    KC_H,    KC_J,    KC_K,    KC_L,    KC_SCLN, KC_QUOT,          KC_ENT,  KC_PGUP, \

--- a/keymap.c
+++ b/keymap.c
@@ -10,6 +10,10 @@ enum alt_keycodes {
     MD_BOOT,               //Restart into bootloader after hold timeout
 };
 
+enum tapdance_keycodes {
+    TD_BACKTICK_ESC,       //Tap Dance Backtick / Esc
+};
+
 const uint16_t PROGMEM keymaps[][MATRIX_ROWS][MATRIX_COLS] = {
     [0] = LAYOUT_65_ansi_blocker(
         KC_GRV,                                                                                                                                \

--- a/keymap.c
+++ b/keymap.c
@@ -31,6 +31,11 @@ void td_backtick_exactly_2_esc(qk_tap_dance_state_t *state, void *user_data) {
     }
 }
 
+// Tap Dance definitions
+qk_tap_dance_action_t tap_dance_actions[] = {
+    [TD_BACKTICK_ESC] = ACTION_TAP_DANCE_FN(td_backtick_exactly_2_esc),
+};
+
 const uint16_t PROGMEM keymaps[][MATRIX_ROWS][MATRIX_COLS] = {
     [0] = LAYOUT_65_ansi_blocker(
         KC_GRV,                                                                                                                                \

--- a/keymap.c
+++ b/keymap.c
@@ -14,6 +14,23 @@ enum tapdance_keycodes {
     TD_BACKTICK_ESC,       //Tap Dance Backtick / Esc
 };
 
+/**
+ * When the key was tapped exactly 2 times, KC_ESC
+ * otherwise, KC_GRV the appropriate number of times.
+ */
+void td_backtick_exactly_2_esc(qk_tap_dance_state_t *state, void *user_data) {
+    int i;
+
+    if (state->count == 2) {
+        tap_code(KC_ESC);
+        return;
+    }
+
+    for (i=0; i<state->count; i++) {
+        tap_code(KC_GRV);
+    }
+}
+
 const uint16_t PROGMEM keymaps[][MATRIX_ROWS][MATRIX_COLS] = {
     [0] = LAYOUT_65_ansi_blocker(
         KC_GRV,                                                                                                                                \

--- a/keymap.c
+++ b/keymap.c
@@ -12,7 +12,8 @@ enum alt_keycodes {
 
 const uint16_t PROGMEM keymaps[][MATRIX_ROWS][MATRIX_COLS] = {
     [0] = LAYOUT_65_ansi_blocker(
-        KC_GRV,  KC_1,    KC_2,    KC_3,    KC_4,    KC_5,    KC_6,    KC_7,    KC_8,    KC_9,    KC_0,    KC_MINS, KC_EQL,  KC_BSPC, KC_DEL,  \
+        KC_GRV,                                                                                                                                \
+                 KC_1,    KC_2,    KC_3,    KC_4,    KC_5,    KC_6,    KC_7,    KC_8,    KC_9,    KC_0,    KC_MINS, KC_EQL,  KC_BSPC, KC_DEL,  \
         KC_TAB,  KC_Q,    KC_W,    KC_E,    KC_R,    KC_T,    KC_Y,    KC_U,    KC_I,    KC_O,    KC_P,    KC_LBRC, KC_RBRC, KC_BSLS, KC_HOME, \
         KC_CAPS, KC_A,    KC_S,    KC_D,    KC_F,    KC_G,    KC_H,    KC_J,    KC_K,    KC_L,    KC_SCLN, KC_QUOT,          KC_ENT,  KC_PGUP, \
         KC_LSFT, KC_Z,    KC_X,    KC_C,    KC_V,    KC_B,    KC_N,    KC_M,    KC_COMM, KC_DOT,  KC_SLSH, KC_RSFT,          KC_UP,   KC_PGDN, \

--- a/rules.mk
+++ b/rules.mk
@@ -1,0 +1,1 @@
+TAP_DANCE_ENABLE = yes

